File: word_chain.rb
require 'set'

class WordChain
  def initialize(n, d)
    @n = n
    @d = d
    @alive = Array.new(n, true)
    @bef = ""
    @p = 0
  end

  def word_chain(s)
    if @d.include?(s) && (@bef.empty? || @bef[-1] == s[0]) && s[-1] != 'z'
      @d.delete(s)
      @bef = s
    else
      @alive[@p] = false
      @bef = ""
    end
    begin
      @p = (@p + 1) % @n
    end while !@alive[@p]
  end

  def alive_count
    @alive.count(true)
  end

  def is_alive(i)
    @alive[i]
  end
end

n, k, m = gets.split.map(&:to_i)
d = Set.new
k.times do
  d.add(gets.chomp)
end

word_chain = WordChain.new(n, d)
m.times do
  word_chain.word_chain(gets.chomp)
end

puts word_chain.alive_count
(0...n).each do |i|
  puts i + 1 if word_chain.is_alive(i)
end

File: word_chain.go
package main

import "fmt"

type WordChain struct {
	N	 int
	D	 map[string]bool
	alive []bool
	bef   string
	P	 int
}

func NewWordChain(n int, d map[string]bool) *WordChain {
	alive := make([]bool, n)
	for i := range alive {
		alive[i] = true
	}
	return &WordChain{
		N:	 n,
		D:	 d,
		alive: alive,
		bef:   "",
		P:	 0,
	}
}

func (wc *WordChain) wordChain(s string) {
	if wc.D[s] && (wc.bef == "" || s[0] == wc.bef[len(wc.bef)-1]) && s[len(s)-1] != 'z' {
		delete(wc.D, s)
		wc.bef = s
	} else {
		wc.alive[wc.P] = false
		wc.bef = ""
	}
	for {
		wc.P = (wc.P + 1) % wc.N
		if wc.alive[wc.P] {
			break
		}
	}
}

func (wc *WordChain) aliveCount() int {
	count := 0
	for _, v := range wc.alive {
		if v {
			count++
		}
	}
	return count
}

func (wc *WordChain) isAlive(i int) bool {
	return wc.alive[i]
}

func main() {
	var N, K, M int
	fmt.Scan(&N, &K, &M)

	D := make(map[string]bool)
	for i := 0; i < K; i++ {
		var d string
		fmt.Scan(&d)
		D[d] = true
	}

	wordChain := NewWordChain(N, D)

	for i := 0; i < M; i++ {
		var s string
		fmt.Scan(&s)
		wordChain.wordChain(s)
	}

	fmt.Println(wordChain.aliveCount())
	for i := 0; i < N; i++ {
		if wordChain.isAlive(i) {
			fmt.Println(i + 1)
		}
	}
}
